### DISCUSSION TITLE
Set DataPerCompany explicitly

### DISCUSSION THREAD

**Discussion by fvet** [StefanMaron/BusinessCentral.LinterCop#37] [New Rule]
It would be helpful to have a code analysis rule that **enforces you to set the DataPerCompany property on a table object explicitly.** Considering this property has a huge impact, it would be good to have the developer be reminded of the choice of making the table a per-company or 'global'/company-wide table. This could initially be a warning that is changed into an error a few major releases later.

https://experience.dynamics.com/ideas/idea/?ideaid=18641d7c-93e4-eb11-ba5e-0003ff45a9bb

_Credits to Rob van Bekkum ;)_

**Comment by StefanMaron** [comment]
You mean, that a warning is raised if a table as no property DataPerCompany set. Resulting in havin the property present on every table either true or false?

**Comment by fvet** [comment]
Indeed, a table - with **TableType = Normal** - should always have the DataPerCompany set.
In our solution, 9 out of 10, the tables are to be defined as **DataPerCompany = false** (while the default = true).

If we would forget to set the property and do not detect this before release (we luckily have scripts in place to do so), this property is impossible to revert. Letting the developer think about it by explicitly setting the DataPerCompany can help us preventing these issues.

**Reply by StefanMaron** [reply]
I see. as this rule would be quite easy to implement I dont not have a problem to do so, but I do not think that this would go along with the main coding guidelines. Most of the projects out there probably dont have even one table table marked as DataPerCompany = false.

Because of this I would set the default severity to Hidden. If anyone needs to have this rule for certain projects it can be changed to a warning or even an error with a custom rule set

**Reply by rvanbekkum** [reply]
It is weird for this particular property not to be a mandatory one while a property like `ApplicationArea` is, which we would set to `All` almost most of the time. The `DataPerCompany` actually has a huge impact if the implicit setting is not what you wanted.

This rule would be really nice. And we'll definitely be including it in our pipeline with at least `warning`-severity. 😊

**Comment by fvet** [comment]
Running version

![image](https://user-images.githubusercontent.com/25268332/134935538-c23bb9c2-1eb9-4195-9cb7-566e946edd8d.png)

seems to fail on LC0007, although it's a temporary table (which should be excluded from the check, not?)

![image](https://user-images.githubusercontent.com/25268332/134935718-63b895fa-54d2-4e7b-a6d2-eda94bb2456d.png)

**Reply by StefanMaron** [reply]
Should be fixed in v0.14
